packages/contracts-bedrock/src/L1/OPContractsManager14.sol

```diff
@@ -0,0 +1,157 @@
+
+            // -------- Discover and Upgrade Proofs Contracts --------
+
+            // All chains have the Permissioned Dispute Game.
+            IPermissionedDisputeGame permissionedDisputeGame = IPermissionedDisputeGame(
+                address(
+                    getGameImplementation(
+                        IDisputeGameFactory(opChainAddrs.disputeGameFactory), GameTypes.PERMISSIONED_CANNON
+                    )
+                )
+            );
+            // We're also going to need the l2ChainId below, so we cache it in the outer scope.
+            uint256 l2ChainId = getL2ChainId(IFaultDisputeGame(address(permissionedDisputeGame)));
+
+            // Now retrieve the permissionless game. If it exists, replace its implementation.
+            IFaultDisputeGame permissionlessDisputeGame = IFaultDisputeGame(
+                address(getGameImplementation(IDisputeGameFactory(opChainAddrs.disputeGameFactory), GameTypes.CANNON))
+            );
+
+            if (address(permissionlessDisputeGame) != address(0)) {
+                // Deploy and set a new permissionless game to update its prestate
+                deployAndSetNewGameImpl({
+                    _l2ChainId: l2ChainId,
+                    _disputeGame: IDisputeGame(address(permissionlessDisputeGame)),
+                    _gameType: GameTypes.CANNON,
+                    _opChainConfig: _opChainConfigs[i],
+                    _implementations: impls,
+                    _blueprints: bps,
+                    _opChainAddrs: opChainAddrs
+                });
+            }
+
+            // Emit the upgraded event with the address of the caller. Since this will be a delegatecall,
+            // the caller will be the value of the ADDRESS opcode.
+            emit Upgraded(l2ChainId, _opChainConfigs[i].systemConfigProxy, address(this));
+        }
+    }
+
+    /// @notice Deploys and sets a new dispute game implementation
+    /// @param _l2ChainId The L2 chain ID
+    /// @param _disputeGame The current dispute game implementation
+    /// @param _gameType The type of game to deploy
+    /// @param _opChainConfig The OP chain configuration
+    /// @param _blueprints The blueprint addresses
+    /// @param _implementations The implementation addresses
+    /// @param _opChainAddrs The OP chain addresses
+    function deployAndSetNewGameImpl(
+        uint256 _l2ChainId,
+        IDisputeGame _disputeGame,
+        GameType _gameType,
+        OpChainConfig memory _opChainConfig,
+        Blueprints memory _blueprints,
+        Implementations memory _implementations,
+        ISystemConfig.Addresses memory _opChainAddrs
+    )
+        internal
+    {
+        // Get the constructor params for the game
+        IFaultDisputeGame.GameConstructorParams memory params =
+            getGameConstructorParams(IFaultDisputeGame(address(_disputeGame)));
+
+        // Set the new vm value.
+        params.vm = IBigStepper(_implementations.mipsImpl);
```

Comment:
To confirm my understanding, this MIPS contract is the shared singleton instance right?

That would mean that after this upgrade every chain would point to the same MIPS contract instance and thus the same PreimageOracle instance. Currently a bunch of chains have their own deployment of MIPS and PreimageOracle which makes monitoring more complex so would be wonderful if that got fixed as part of this upgrade.  Existing games would keep pointing to whatever MIPS/PreimageOracle they were using previously so there's no compatibility concerns here, I'm just checking my understanding to ensure we really do get everything switched to using the singletons.

Reply:
Yes - this is my understanding.  @maurelian or @mslipper - can you confirm?